test/e2e/test_logging.go

```diff
@@ -0,0 +1,266 @@
+/*
+Copyright 2020 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"testing"
+	"time"
+
+	duckv1 "github.com/google/knative-gcp/pkg/apis/duck/v1"
+	sourcev1 "github.com/google/knative-gcp/pkg/apis/events/v1"
+	v1 "github.com/google/knative-gcp/pkg/apis/intevents/v1"
+	"github.com/google/knative-gcp/test/lib"
+	"github.com/google/uuid"
+	cloudlogging "google.golang.org/api/logging/v2"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"knative.dev/eventing/test/lib/duck"
+	"knative.dev/eventing/test/lib/resources"
+	"knative.dev/pkg/apis"
+	pkgduckv1 "knative.dev/pkg/apis/duck/v1"
+	"knative.dev/pkg/test/helpers"
+
+	// The following line to load the gcp plugin (only required to authenticate against GKE clusters).
+	_ "k8s.io/client-go/plugin/pkg/client/auth/gcp"
+)
+
+const (
+	iso8601 = "2006-01-02T15:04:05.999Z"
+)
+
+var (
+	trueVal  = true
+	falseVal = false
+)
+
+func CloudLoggingGCPControlPlaneTestImpl(t *testing.T, authConfig lib.AuthConfig) {
+	ctx := context.Background()
+	startTimestamp := time.Now()
+
+	client := lib.Setup(ctx, t, true, authConfig.WorkloadIdentity)
+	defer lib.TearDown(ctx, client)
+
+	topicName := helpers.AppendRandomString("test-e2e-gcp-control-plane-")
+	randomString := uuid.New().String()
+
+	topic := &v1.Topic{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "internal.events.cloud.google.com/v1",
+			Kind:       "Topic",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "cloud-logging-test-topic",
+			Annotations: map[string]string{
+				// This will cause the controller and webhook to write randomString to their logs,
+				// which we will check for using the Cloud Logging API below.
+				v1.LoggingE2ETestAnnotation: randomString,
+			},
+		},
+		Spec: v1.TopicSpec{
+			Topic:           topicName,
+			EnablePublisher: &falseVal,
+		},
+	}
+	client.CreateTopicOrFail(topic)
+	err := duck.WaitForResourceReady(client.Core.Dynamic, &resources.MetaResource{
+		TypeMeta:   topic.TypeMeta,
+		ObjectMeta: topic.ObjectMeta,
+	})
+	if err != nil {
+		t.Fatalf("Topic did not become ready: %v", err)
+	}
+
+	// Sleep so that the logging API has a chance to receive the logs lines and process them before
+	// we try to retrieve them.
+	time.Sleep(2 * time.Minute)
```

Comment:
I understand we have no other way but to wait here. but to my experience magic lines like this are usually a source of flakiness.

have you considered putting this part in a loop to sleep and retrying a few times? that will help reduce flakiness, while avoiding long waits in the test.

Reply:
Turned it into a loop.